src/main/java/net/helix/pendulum/crypto/Merkle.java

```diff
@@ -55,6 +60,44 @@
     }
 
 
+    public static List<byte[]> buildMerkleTransactionTree(List<Hash> leaves, Hash milestoneHash, Hash address){
+        if (leaves.isEmpty()) {
+            leaves.add(Hash.NULL_HASH);
+        }
+        byte[] buffer;
+        Sponge sha3 = SpongeFactory.create(SpongeFactory.Mode.S256);
+        int depth = (int) Math.ceil(Math.sqrt(leaves.size()));
+        int row = 1;
+        List<byte[]> virtualTransactionList = new ArrayList<byte[]>();
+        while (leaves.size() > 1) {
+            List<Hash> nextKeys = Arrays.asList(new Hash[(leaves.size() / 2)]);
+            for (int i = 0; i < nextKeys.size(); i++) {
+                if (areLeavesNull(leaves, i)) continue;
+                sha3.reset();
+                Hash k1 = leaves.get(i * 2);
+                Hash k2 = leaves.get(i * 2 + 1);
+                buffer = computeParentHash(sha3, k1, k2);
+                Hash parentHash = HashFactory.TRANSACTION.create(buffer);
+                nextKeys.set(i, parentHash);
+               // if( i == 0) add only one virtual transaction for each level
+                virtualTransactionList.add(createVirtualTransaction(k1, k2, row, milestoneHash, address, parentHash));
+            }
+            leaves = nextKeys;
+            row++;
+        }
+        return virtualTransactionList;
+    }
+
+    private static byte[] computeParentHash(Sponge sha3, Hash k1, Hash k2) {
+        byte[] buffer;
+        buffer = Arrays.copyOfRange(k1 == null ? Hex.decode("0000000000000000000000000000000000000000000000000000000000000000") : k1.bytes(), 0, 32);
+        sha3.absorb(buffer, 0, buffer.length);
+        buffer = Arrays.copyOfRange(k2 == null ? Hex.decode("0000000000000000000000000000000000000000000000000000000000000000") : k2.bytes(), 0, 32);
+        sha3.absorb(buffer, 0, buffer.length);
+        sha3.squeeze(buffer, 0, buffer.length);
+        return buffer;
+    }
+
     public static List<List<Hash>> buildMerkleTree(List<Hash> leaves){
```

Comment:
What are the pros and cons of having `List<List<Hash>>` for the tree representation against a flat array (`ArrayList<Hash>`) ?

Reply:
It depends on what we want to use it:
List<List<Hash>> is faster if you need to access hashes for a certain level of Merkle tree, but it's slower on traverse all nodes (without any order), or traverse from leves to root.
List<Hash> is faster if all nodes should be traversed and if merkle level is not relevant
In most cases I saw that only root is used, so maybe we can change the calling methods to call getMerkleRoot instead of the entire tree.

Reply:
The trick is to use the compact array representation of a binary tree. If you have say 8 leaves, you pack it into array of size `15 = (2*8) - 1`  in the following way: 
- leaves have indices 7, 8, ..., 14
- next level indices 3,4,5,6
- next level 1,2
- root 0

The level `i` has indices from `2^i-1` to `2^{i+1}-2` 
The benefit of having this representation is that you initially allocate the full size and fill  evertyhing with `null` and as the data becomes available, you can quickly navigate to parents and children by simple bit shifts of the indices, at the same time preserving the order at all levels.

Reply:
To be more precise: children of `a[i]` are `a[2*i+1]` and `a[2*i+2]`, parent of  `a[i]` is `a[(i-1) >> 1]`

Reply:
I'd suggest introducing a separate interface MerkleTree and MerkleTreeImpl, where this builder can be moved.

Also, I don't think that we should replace `null` with `NULL_HASH` at all levels, it should be used *only* for padding the leaves list so that it's size is an integer of the form `2^i` for some `i`.